qermit/coherent_pauli_checks/box_clifford_subcircuits.py

```diff
@@ -0,0 +1,316 @@
+
+    if command.op.type == OpType.Rz:
+        if command.op.params == [0.5]:
+            return True
+
+    if command.op.type == OpType.PhasedX:
+        if command.op.params == [0.5, 0.5]:
+            return True
+
+    return False
+
+
+def _get_clifford_commands(command_list: list[Command]) -> set[int]:
+    """Given a list of commands, return a set of indexes of that list
+    corresponding to those commands which are Clifford gates.
+
+    :param command_list: List of commands in which to search for
+        Clifford commends.
+    :return: Indexes in the list which correspond to commands in the
+        list which are Clifford.
+    """
+    return {
+        i for i, command in enumerate(command_list) if _command_is_clifford(command)
+    }
+
+
+def _give_nodes_subdag(dag: nx.DiGraph, node_subdag: dict[int, int]) -> list[int]:
+    """Assign a sub-DAG to all nodes in given dag. Some may already have
+    an assigned sub-DAG as given and these are preserved. Nodes without an
+    assigned sub-DAG are given a unique sub-DAG of their own.
+
+    :param dag: Directed acyclic graph.
+    :param node_subdag: Map from node to sub-DAG for those nodes with
+        an existing assignment.
+    :raises Exception: Raised if nodes are not sequential integers.
+    :return: List of sub-DAGs. List is indexed by node.
+    """
+
+    if not sorted(list(dag.nodes)) == [i for i in range(dag.number_of_nodes())]:
+        raise Exception("The nodes of the given dag must be sequential integers.")
+
+    node_subdag_list = []
+    subdag_index = 0
+    for node in range(dag.number_of_nodes()):
+        if node in node_subdag.keys():
+            node_subdag_list.append(node_subdag[node])
+        else:
+            while (subdag_index in node_subdag.values()) or (
+                subdag_index in node_subdag_list
+            ):
+                subdag_index += 1
+            node_subdag_list.append(subdag_index)
+
+    return node_subdag_list
+
+
+def _circuit_to_graph(circuit: Circuit) -> tuple[nx.DiGraph, list[Command]]:
+    """Convert circuit to graph. Nodes correspond to commands,
+    edges indicate a dependence between the outputs and inputs of
+    two commands. Node values corresponds to indexes in the returned
+    list of commands.
+
+    :param circuit: Circuit to convert to a graph.
+    :return: Tuple of graph and list of commands. Nodes are indexes
+        in the list of commands.
+    """
+    # Lists the most recent node to act on a particular qubits. If a
+    # new gate is found to act on that qubit then an edge between the
+    # node which corresponds to the new gate and the node which
+    # most recently acted on that qubit is added. This builds a DAG
+    # showing the temporal order of gates.
+    node_command = circuit.get_commands()
+    current_node: dict[Qubit, int] = {}
+    dag = nx.DiGraph()
+    for node, command in enumerate(node_command):
+        dag.add_node(node)
+        for qubit in command.qubits:
```

Comment:
Is it assumed that this is a pure quantum circuit with no mid-circuit measurements?

Reply:
Clifford circuits are only those that are pure quantum gates. There can be gates which are mixed quantum and classical, but they will always be treated as non Clifford, and in particular not coloured in the graph.